lib/Parse/ParseType.cpp

```diff
@@ -392,47 +399,58 @@ ParserResult<TypeRepr> Parser::parseTypeIdentifier() {
   return makeParserResult(Status, ITR);
 }
 
-/// parseTypeIdentifierOrTypeComposition
-/// - Identifiers and compositions both start with the same identifier
-///   token, parse it and continue constructing a composition if the
-///   next token is '&'
+ParserResult<TypeRepr> Parser::parseTypeSimpleOrComposition() {
+  return parseTypeSimpleOrComposition(diag::expected_identifier_for_type);
+}
+
+/// parseTypeSimpleOrComposition
 ///
 ///   type-composition:
-///     type-identifier ('&' type-identifier)*
-///     type-composition-deprecated
-///
-///   type-composition-list-deprecated:
-///     type-identifier (',' type-identifier)*
-ParserResult<TypeRepr> Parser::parseTypeIdentifierOrTypeComposition() {
-  if (Tok.is(tok::kw_protocol) && startsWithLess(peekToken()))
-    return parseOldStyleProtocolComposition();
-
-  SourceLoc FirstTypeLoc = Tok.getLoc();
-  
+///     type-simple
+///     type-composition '&' type-simple
+ParserResult<TypeRepr> Parser::parseTypeSimpleOrComposition(Diag<> MessageID,
+                                                    bool HandleCodeCompletion) {
```

Comment:
Style nitpick: we try to avoid right-aligning things. One option is to break after the return type, but if that's not good enough it's better to break after the open-paren and just double-indent the parameters.